scipy/stats/_distribution_infrastructure.py

```diff
@@ -3618,62 +3682,99 @@ def _cdf2_subtraction(self, x, y, **params):
-        res = np.where(res.fl >= 0, np.floor(res.xl), np.floor(res.xr))
+        # follows same logic as _icdf_inversion
+        res = self._solve_bounded(self._logcdf_dispatch, x, params=params, xatol=0.9)
+        res = np.where(res.fun >= 0, np.floor(res.x), np.ceil(res.x))
+        res = np.where(self.logcdf(res) >= x, res, res + 1.0)
         res[np.isnan(x)] = np.nan
         return res
 
     def _iccdf_inversion(self, x, **params):
-        res = self._solve_bounded(self._ccdf_dispatch, x, params=params, xatol=1)
-        res = np.where(res.fl <= 0, np.floor(res.xl), np.floor(res.xr))
+        res = self._solve_bounded(self._ccdf_dispatch, x, params=params, xatol=0.9)
+        res = np.where(res.fun <= 0, np.floor(res.x), np.ceil(res.x))
+        res = np.where(self.ccdf(res) <= x, res, res + 1.0)
         res[np.isnan(x)] = np.nan
         return res
 
     def _ilogccdf_inversion(self, x, **params):
-        res = self._solve_bounded(self._logccdf_dispatch, x, params=params, xatol=1)
-        res = np.where(res.fl <= 0, np.floor(res.xl), np.floor(res.xr))
+        res = self._solve_bounded(self._logccdf_dispatch, x, params=params, xatol=0.9)
+        res = np.where(res.fun <= 0, np.floor(res.x), np.ceil(res.x))
+        res = np.where(self.logccdf(res) <= x, res, res + 1.0)
         res[np.isnan(x)] = np.nan
         return res
 
     def _mode_optimization(self, **params):
         # If `x` is the true mode of a unimodal continuous function, we can find
-        # the mode among the integers by rounding in each direction and checking
-        # which is better. I think `xatol` should be able to be 1 if the documented
-        # convergence criterion were implemented, but the actual convergence
-        # criterion is different, and I'm not sure if it controls the total width
-        # of the bracket. `xatol=0.1` might not be safe as-implemented.
-        x = super()._mode_optimization(xatol=0.1, **params)
+        # the mode among integers by rounding in each direction and checking
+        # which is better. If the difference between `x` and the nearest integer
+        # is less than `xatol`, the computed value of `x` may end up on the wrong
+        # side of the nearest integer. Setting `xatol=0.5` guarantees that at most
+        # three integers need to be checked, the two nearest integers, ``floor(x)``
+        # and ``round(x)`` and the nearest integer other than these.
+        x = super()._mode_optimization(xatol=0.5, **params)
+        low, high = self.support()
         xl, xr = np.floor(x), np.ceil(x)
-        fl, fr = self._pmf_dispatch(xl, **params), self._pmf_dispatch(xr, **params)
+        nearest = np.round(x)
+        # Clip to stay within support. There will be redundant calculation
+        # when clipping since `xo` will be one of `xl` or `xr`, but let's
+        # keep the implementation simple for now.
+        xo = np.clip(nearest + np.copysign(1, nearest - x), low, high)
```

Comment:
I think a little simpler than the `copysign` stuff is if you use `xl` or `xr` as the basis for `floor` and `round`. ISTM that if you use `xr`, for instance, you need to `floor` and `ceil` that, and also look at `floor - 1`. Right?

Reply:
I ended up stacking the three values and vectorizing, but kept the other parts the same. I think it's simpler to keep like as is for now because

1. It's easier to call `super()._mode_optimization` rather than to use `_chandrupatla` directly to get `xl` and `xr`.
2. `xl` and `xr` are bracket endpoints, and won't be `nan` when the result should be `nan` so we still need to look at something else to figure out if the result should be `nan`, (like I did above in `_base_discrete_inversion`).

That's a good point that if `xatol = 0.5` then the three points to check are guaranteed to be `ceil(xr)`, `floor(xr)`, and `floor(xr)  - 1`. That sounds right. I suggest revisiting this later, and prioritizing getting discrete distributions in, since this code is at least working.

Reply:
OK.